### PR TITLE
[SPARK-9929][SQL] support metadata in withColumn

### PR DESCRIPTION
**Pull request by cloud-fan**
in MLlib sometimes we need to set metadata for the new column, thus we will alias the new column with metadata before call `withColumn` and in `withColumn` we alias this clolumn again. Here I overloaded `withColumn` to allow user set metadata, just like what we did  for `Column.as`.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @rxin , this blocks  https://github.com/apache/spark/pull/7957

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -1150,6 +1150,25 @@ class DataFrame private[sql](
   }
 
   /**
+   * Returns a new [[DataFrame]] by adding a column with medadata.
```
Comment (jkbradley) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: typo: metadata

**Comment by jkbradley** [comment]
The ML changes look good.  (Thanks for adding this.)

The new method looks fine, unless you want to reduce code duplication.

**Comment by rxin** [comment]
Does the new function needs to be public? Can it be private[spark] ?

**Comment by jkbradley** [comment]
ML does not need it to be public.

**Comment by rxin** [comment]
OK @cloud-fan let's make this private[spark] for now.

**Comment by rxin** [comment]
The API change LGTM.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 2]
```diff
@@ -1150,6 +1150,23 @@ class DataFrame private[sql](
   }
 
   /**
+   * Returns a new [[DataFrame]] by adding a column with metadata.
+   */
+  private[spark] def withColumn(colName: String, col: Column, metadata: Metadata): DataFrame = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: should we make `Column.as(alias: String, metadata: Metadata)` also private? As it also expose the `Metadata` too.

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Don't think we can change that since it's been public since 1.3.

**Comment by jkbradley** [comment]
LGTM

**Comment by rxin** [comment]
Merging this in master (not branch-1.5).

**Pull request closed without merging**